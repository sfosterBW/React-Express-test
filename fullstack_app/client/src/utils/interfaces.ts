//Objects
export interface IFruit {
  _id: number
  name: string
  best: boolean
}

export interface IFruitResponse {
  status: number
}

//States
export interface AlertState {
  toggle: boolean
}

export interface ModalState {
  toggle: boolean
}
