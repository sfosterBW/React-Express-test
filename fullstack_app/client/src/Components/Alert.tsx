import React, { FC } from 'react'

interface Props {
  onClose: () => void
  title: string
}

const Alert: FC<Props> = ({ onClose, title }) => {
  return (
    <div>
      <p>{title}</p>
      <button onClick={onClose}>&times;</button>
    </div>
  )
}

export default Alert
