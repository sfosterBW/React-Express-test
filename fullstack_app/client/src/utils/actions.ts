//Action const
export const TOGGLE_ALERT = "TOGGLE_ALERT"

//Action shape
interface ToggleAlert {
  type: typeof TOGGLE_ALERT,
  payload: boolean
}

//Action type
export type AlertActionTypes = ToggleAlert

//Action
export function toggleAlert(alertValue: boolean): AlertActionTypes {
  return {
    type: TOGGLE_ALERT,
    payload: alertValue
  }
}

//Action const
export const TOGGLE_MODAL = "TOGGLE_MODAL"

//Action shape
interface ToggleModal {
  type: typeof TOGGLE_MODAL,
  payload: boolean
}

//Action type
export type ModalActionTypes = ToggleModal

//Action
export function toggleModal(modalValue: boolean): ModalActionTypes {
  return {
    type: TOGGLE_MODAL,
    payload: modalValue
  }
}
