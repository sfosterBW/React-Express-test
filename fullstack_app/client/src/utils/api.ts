import axios from 'axios'
import { IFruit } from './interfaces'
 
axios.defaults.baseURL = '/fruit-api/'

const fetchFruit = async () => {
  const response = await axios.get("list")
  return response.data
}

const createFruit = async (fruit: { name: string, best: boolean }) => {
    const response = await axios.post('new', { new: fruit })
    return response.data
}

const updateFruit = async (fruit: IFruit) => {
    const response = await axios.put('update', fruit)
    return response.data
}

const deleteFruit = async (id: number) => {
  const deleteData = { params: { id: id } }
  const response = await axios.delete('delete', deleteData)
  return response.data
}

export default { fetchFruit, createFruit, updateFruit, deleteFruit }
