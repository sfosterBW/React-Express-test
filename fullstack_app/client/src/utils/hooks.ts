import { useState } from 'react'

export const useField = (
  initValue: string,
  label: string,
  name: string,
  type: string) => {

  const [value, setValue] = useState(initValue)

  const handleChange = (event: any) => {
    setValue(event.target.value)
  }

  const reset = () => {
    setValue("")
  }

  return {
    handleChange,
    label,
    name,
    reset,
    type,
    value
  }
}
