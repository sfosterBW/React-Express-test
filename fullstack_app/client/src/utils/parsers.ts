import { Fruit } from './interfaces'

const isBoolean = (bool: any): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: any): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: any): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

export const parseId = (id: any): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: any): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseArray = (array: any): any[] => {
  if(array && !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${array}`);
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  return {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
