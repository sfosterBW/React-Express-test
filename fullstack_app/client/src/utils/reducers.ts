import { AlertState, FruitState, ModalState } from './interfaces'
import {
  TOGGLE_ALERT,
  AlertActionTypes,
  TOGGLE_MODAL,
  OPEN_MODAL,
  ModalActionTypes,
  GET_FRUITS,
  CREATE_FRUIT,
  REMOVE_FRUIT,
  UPDATE_FRUIT,
  FruitActionTypes,
} from './actions'

export const initialAlertState: AlertState = {
  toggle: false
}

export function alertReducer(
  state = initialAlertState,
  action: AlertActionTypes): AlertState {
  switch (action.type) {
    case TOGGLE_ALERT: {
      return { ...state, toggle: action.payload || false }
    }
    default:
      return { ...state }
  }
}

export const initialModalState: ModalState = {
  toggle: false,

}

export function modalReducer(
  state = initialModalState,
  action: ModalActionTypes): ModalState {
  switch (action.type) {
    case TOGGLE_MODAL: {
      return { ...state, toggle: action.payload || false }
    }
    case OPEN_MODAL: {
      return { toggle: true, fruit: action.payload }
    }
    default:
      return { ...state }
  }
}

export const initialFruitState: FruitState = {
  data: []
}

export function fruitReducer(
  state = initialFruitState,
  action: FruitActionTypes
): FruitState {
  switch (action.type) {
    case GET_FRUITS: {
      return { ...state, data: action.payload }
    }
    case CREATE_FRUIT: {
      return { ...state, data: state.data.concat(action.payload) }
    }
    case REMOVE_FRUIT: {
      const newFruits = state.data.filter(i => i.id !== action.payload)
      return { ...state, data: newFruits }
    }
    case UPDATE_FRUIT: {
      const updateFruit = state.data.map((fruit) => {
        if (fruit.id === action.payload.id) {
          fruit.best = action.payload.best
          fruit.name = action.payload.name
          return fruit
        } else {
          return fruit
        }
      })
      return { ...state, data: updateFruit }
    }
    default:
      return { ...state }
  }
}
