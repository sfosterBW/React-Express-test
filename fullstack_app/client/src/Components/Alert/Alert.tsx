import React, { FC } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { RootState } from '../../utils/store'
import { toggleAlert } from '../../utils/actions'
import styles from './Alert.module.css'

const Alert: FC = () => {
  const selectToggle = (state: RootState) => state.alert.toggle
  const toggle = useSelector(selectToggle)
  const dispatch = useDispatch()

  if(!toggle) {
    return null
  }

  return (
    <div className={styles.wrapper}>
      <section className={styles.alert}>
        <p className={styles.message}>{String(toggle)}</p>
        <button
          className={styles.button}
          onClick={() => {dispatch(toggleAlert(!toggle))}}
        >
          &times;
        </button>
      </section>
    </div>
  )
}

export default Alert
