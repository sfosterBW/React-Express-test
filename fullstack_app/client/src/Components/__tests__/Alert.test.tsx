import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { createStore } from 'redux'

import Alert from '../Alert'
import { rootReducer } from '../../utils/reducers'

describe('the Alert component', () => {

  const value = true
  const mockStore = createStore(rootReducer, { value: value })
  const alertComponent =
    <Provider store={mockStore}>
      <Alert />
    </Provider>
  const alert = mount(alertComponent)

  it('renders with the right structure', () => {
    expect(alert).toBeDefined()
    expect(alert.find('div')).toHaveLength(2)
    expect(alert.find('p')).toHaveLength(1)
    expect(alert.find('p').text()).toBe(String(value))
    expect(alert.find('button')).toHaveLength(1)
  })

  it('renders on mount functions as expected', () => {
    alert.find('button').simulate('click')
    expect(alert.find('div')).toHaveLength(1)
  })

  it('renders the same as last time with props', () => {
    const tree = renderer
      .create(alertComponent)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
