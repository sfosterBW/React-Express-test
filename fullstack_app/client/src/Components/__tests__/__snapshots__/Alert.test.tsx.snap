// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Alert component renders the same as last time with props 1`] = `<div />`;
